File: python/1590. Make Sum Divisible by P.py
class Solution(object):
    def minSubarray(self, nums, p):
        """
        :type nums: List[int]
        :type p: int
        :rtype: int
        """
        n = len(nums)
        s = sum(nums)
        if s%p==0:
            return 0
        k = s%p
        mem = {0:0}
        s = 0
        ans = n
        for i in range(1, n+1):
            s += nums[i-1]
            x = (s - k + p)%p
            if x in mem:
                ans = min(ans, i - mem[x])
            mem[s%p] = i
        return ans if ans != n else -1

# 思路：寻找最短的子数组，满足子数组的和s%p=k, 其中k=sum(nums)%p
# t = (a - k)%p, b%p == t. --> (a-b)%p = k

File: go/src/leetcode/1590/1590. Make Sum Divisible by P.go
package main

import "fmt"

/*
*
思路: nums的和为s, k = s%p
子数组的和为a, 要求a%p==k, 则(s -a)%p == s%p - a%p = 0,满足去掉a之后，剩余元素和(mod p) == 0
如何求满足a%p==k的子数组？
通过前缀和, (a-k)%p==0,
*/
func minSubarray(nums []int, p int) int {
	n := len(nums)
	var ans = n
	var mem = map[int]int{0: 0}
	var s int
	for _, x := range nums {
		s = (s + x) % p
	}
	if s%p == 0 {
		return 0
	}

	k := s % p
	s = 0
	for i := 1; i <= n; i++ {
		s = (s + nums[i-1]) % p
		if j, ok := mem[(s-k+p)%p]; ok {
			ans = min(ans, i-j)
		}
		mem[s] = i
	}

	if ans == n {
		return -1
	}
	return ans
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	fmt.Println(minSubarray([]int{3, 1, 4, 2}, 6))
	fmt.Println(minSubarray([]int{6, 3, 5, 2}, 9))
	fmt.Println(minSubarray([]int{1, 2, 3}, 3))
	fmt.Println(minSubarray([]int{1, 2, 3}, 7))
	fmt.Println(minSubarray([]int{1000000000, 1000000000, 1000000000}, 3))
}
